In [1]:
import os, sys
if os.path.basename(os.getcwd()) == "notebooks":
    os.chdir("..")                      # run everything from the project root
sys.path.insert(0, os.getcwd())

import warnings; warnings.filterwarnings("ignore")
import numpy as np, pandas as pd
import matplotlib.pyplot as plt, seaborn as sns
%matplotlib inline
pd.set_option("display.max_columns", 50)

# 02 · Data Cleaning
We apply `src.data_preprocessing.clean_data` and save `data/processed/visitors_cleaned.csv`.

In [2]:
from src.data_preprocessing import load_raw, clean_data, save_processed

raw = load_raw()
print('raw     :', raw.shape)
print('dups    :', raw.duplicated().sum())
print('months  :', raw.Month.unique())

raw     : (12330, 18)
dups    : 125
months  : <ArrowStringArray>
['Feb', 'Mar', 'May', 'Oct', 'June', 'Jul', 'Aug', 'Nov', 'Sep', 'Dec']
Length: 10, dtype: str


In [3]:
clean = clean_data(raw)
print('cleaned :', clean.shape)
print('months  :', clean.Month.unique())
print('missing :', clean.isna().sum().sum())
clean.head()

cleaned : (12205, 19)
months  : <ArrowStringArray>
['Feb', 'Mar', 'May', 'Oct', 'Jun', 'Jul', 'Aug', 'Nov', 'Sep', 'Dec']
Length: 10, dtype: str
missing : 0


,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration,BounceRates,ExitRates,PageValues,SpecialDay,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue,Month_Num
0,0,0.0,0,0.0,1,0.000000,0.20,0.20,0.0,0.0,Feb,1,1,1,1,Returning_Visitor,0,0,2
1,0,0.0,0,0.0,2,64.000000,0.00,0.10,0.0,0.0,Feb,2,2,1,2,Returning_Visitor,0,0,2
2,0,0.0,0,0.0,1,0.000000,0.20,0.20,0.0,0.0,Feb,4,1,9,3,Returning_Visitor,0,0,2
3,0,0.0,0,0.0,2,2.666667,0.05,0.14,0.0,0.0,Feb,3,2,2,4,Returning_Visitor,0,0,2
4,0,0.0,0,0.0,10,627.500000,0.02,0.05,0.0,0.0,Feb,3,3,1,4,Returning_Visitor,1,0,2


In [4]:
assert (clean.select_dtypes('number') >= 0).all().all()
assert clean[['BounceRates','ExitRates']].max().max() <= 1
print('Sanity checks passed')

Sanity checks passed


In [5]:
path = save_processed(clean)
print('Saved to', path)

Saved to data/processed/visitors_cleaned.csv


**What changed:** 125 duplicate sessions dropped, `June`→`Jun`, numeric `Month_Num` added, `Weekend`/`Revenue` converted to 0/1, plus defensive imputation and clipping (no-ops on this file, but protect against future data).